### PR TITLE
✨🤖 add ansible tasks to create a k8s cluster on vsphere

### PR DESCRIPTION
**Pull request by bmiddha**
# Deploy RKE on vSphere
- Added ansible tasks to create a k8s cluster using RKE on vSphere.

### CODE REVIEW

File: ansible-k8s-vsphere/templates/user-data.j2 [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      "log-opts": {
+        "max-size": "100m"
+      },
+      "storage-driver": "overlay2"
+    }
+  path: /etc/docker/daemon.json
+
+runcmd:
+# Install Docker
+- curl -fsSL https://get.docker.com | sh
```
Comment (clee231) [inline ansible-k8s-vsphere/templates/user-data.j2:43]: From: https://docs.docker.com/engine/install/ubuntu/#install-using-the-convenience-script
> Docker provides convenience scripts at get.docker.com and test.docker.com for installing edge and testing versions of Docker Engine - Community into development environments quickly and non-interactively. The source code for the scripts is in the docker-install repository. **Using these scripts is not recommended for production environments, and you should understand the potential risks before you use them**

In the future, I think we can switch this out to use the recommended installation method of using the docker apt repositories. ([Installation Methods](https://docs.docker.com/engine/install/ubuntu/#installation-methods))

File: ansible-k8s-vsphere/templates/user-data.j2 [hunk 2]
```diff
@@ -0,0 +1,57 @@
+#cloud-config
```
Comment (clee231) [inline ansible-k8s-vsphere/templates/user-data.j2:1]: :heart: Love the use of cloud-init here! :+1:

File: ansible-k8s-vsphere/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    file:
+      path: "{{ rke_bin }}"
+      mode: 0755
+
+  - name: Genereate cluster config from template to {{ work_dir }}/{{ cluster_name }}.yml
+    template:
+      src: ../templates/cluster.j2
+      dest: "{{ work_dir }}/{{ cluster_name }}.yml"
+
+  - name: Initite wait for VM IPs
```
Comment (clee231) [inline ansible-k8s-vsphere/tasks/main.yml:402]:
```suggestion
  - name: Initiate wait for VM IPs
```

File: ansible-k8s-vsphere/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+  - name: Set rke path (rke_bin) variable
+    set_fact:
+      rke_bin: "{{ work_dir }}/rke_{{ rke_version }}"
+
+  - name: Ensure rke executable
+    file:
+      path: "{{ rke_bin }}"
+      mode: 0755
+
+  - name: Genereate cluster config from template to {{ work_dir }}/{{ cluster_name }}.yml
```
Comment (clee231) [inline ansible-k8s-vsphere/tasks/main.yml:397]:
```suggestion
  - name: Generate cluster config from template to {{ work_dir }}/{{ cluster_name }}.yml
```

File: ansible-k8s-vsphere/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+      cluster_nodes: "{{ cluster_nodes | default([]) + [item] }}"
+    with_sequence: "start={{ 1 }} end={{ cluster_num_nodes }} stride=1 format={{ node_name }}"
+
+  - name: Build cluster control plane nodes (cluster_nodes_control_plane) array
+    set_fact: 
+      cluster_nodes_control_plane: "{{ cluster_nodes_control_plane | default([]) + [item] }}"
+    with_sequence: "start={{ 1 }} end={{ cluster_num_control_plane }} stride=1 format={{ node_name }}"
+
+  - block:
+    - name: Bandaid
```
Comment (clee231) [inline ansible-k8s-vsphere/tasks/main.yml:66]: :adhesive_bandage: This is for your inventory edge-case, right? Where it complains about the last node, right?

File: ansible-k8s-vsphere/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,426 @@
+- name: test vsphere
```
Comment (clee231) [inline ansible-k8s-vsphere/tasks/main.yml:1]:
```suggestion
- name: Deploy RKE on vSphere
```

**Review by clee231 — CHANGES_REQUESTED** [review]
Just a few typo changes. 
I noted some other changes that could take place later. The larger changes don't need to happen as part of this PR.

File: ansible-k8s-vsphere/tasks/main.yml (hunk 5 above)
Comment (bmiddha) [reply ansible-k8s-vsphere/tasks/main.yml:66]: Yes. I think `with_sequence` is evaluated before `when`. It complains it since `start` > `end` when `cluster_num_control_plane == cluster_num_nodes `


```yml
    - name: Build cluster worker nodes (cluster_nodes_worker) array
      set_fact: 
        cluster_nodes_worker: "{{ cluster_nodes_worker | default([]) + [item] }}"
      with_sequence: "start={{ cluster_num_control_plane + 1}} end={{ cluster_num_nodes }} stride=1 format={{ node_name }}"
      when: "{{ cluster_num_nodes > cluster_num_control_plane }}"
```

**Review by clee231 — APPROVED** [review]

File: ansible-k8s-vsphere/templates/user-data.j2 (hunk 1 above)
Comment (bmiddha) [reply ansible-k8s-vsphere/templates/user-data.j2]: resolved in 63f0e5db57a904e982f4bfac7686a9256d6340c7
